src/utils/AdaptiveTimeStep.cpp

```diff
@@ -239,6 +232,67 @@ AdaptiveTimeStep::Calculate (
     }
 }
 
+void
+AdaptiveTimeStep::GatherMinAcc (MultiBeam& beams, const amrex::Geometry& geom,
+                                const Fields& fields, const int it)
+{
+    HIPACE_PROFILE("AdaptiveTimeStep::Calculate()");
+    using namespace amrex::literals;
+
+    if (m_do_adaptive_time_step == 0) return;
+
+    const PhysConst phys_const = get_phys_const();
+    const amrex::Real clightinv = 1._rt/phys_const.c;
+
+    constexpr int lev = 0;
+
+    // Extract properties associated with physical size of the box
+    const int nbeams = beams.get_nbeams();
+    const int numprocs_z = Hipace::GetInstance().m_numprocs_z;
+
+    for (int ibeam = 0; ibeam < nbeams; ibeam++) {
+        if (!Hipace::HeadRank()) break;
+        const auto& beam = beams.getBeam(ibeam);
+        const amrex::Real charge_mass_ratio = beam.m_charge / beam.m_mass;
+
+        const uint64_t box_offset = beam.m_box_sorter.boxOffsetsPtr()[beam.m_ibox];
+        const uint64_t numParticleOnTile = beam.m_box_sorter.boxCountsPtr()[beam.m_ibox];
+
+        // Extract particle properties
+        const auto& aos = beam.GetArrayOfStructs(); // For positions
+        const auto& pos_structs = aos.begin() + box_offset;
+
+        amrex::ReduceOps<amrex::ReduceOpMin> reduce_op;
+        amrex::ReduceData<amrex::Real> reduce_data(reduce_op);
+        using ReduceTuple = typename decltype(reduce_data)::Type;
+
+        // Data required to gather the Ez field
+        const amrex::FArrayBox& slice_fab = fields.getSlices(lev)[0];
+        Array3<const amrex::Real> const slice_arr = slice_fab.const_array();
+        const int ez_comp = Comps[WhichSlice::This]["Ez"];
+        const amrex::Real dx_inv = geom.InvCellSize(0);
+        const amrex::Real dy_inv = geom.InvCellSize(1);
+        amrex::Real const x_pos_offset = GetPosOffset(0, geom, slice_fab.box());
+        const amrex::Real y_pos_offset = GetPosOffset(1, geom, slice_fab.box());
+
+        reduce_op.eval(numParticleOnTile, reduce_data,
+            [=] AMREX_GPU_DEVICE (long ip) noexcept -> ReduceTuple
+            {
+                if (pos_structs[ip].id() < 0) return { 0._rt };
```

Comment:
Thanks, I fixed the binning handling, and indeed we operate on `This` (although this would not mame much of a difference, this is really a heuristic).